datafusion/functions/src/encoding/meta.rs

```diff
@@ -0,0 +1,108 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Metadata information for "encode" and "decode" functions
+use datafusion_common::arrow::datatypes::DataType;
+use datafusion_common::{plan_err, DataFusionError, Result};
+use datafusion_expr::TypeSignature::*;
+use datafusion_expr::{ColumnarValue, FunctionImplementation, Signature, Volatility};
+use std::sync::OnceLock;
+use DataType::*;
+
+pub(super) struct EncodeFunc {}
+
+static ENCODE_SIGNATURE: OnceLock<Signature> = OnceLock::new();
```

Comment:
Here is what `encode` and `decode` look like using the `ScalarUDF` API -- I think they are much clearer when all this type information is in one place (though I still kept it separate from the implementation to show the implementation did not change at all)